### PR TITLE
Add life cycle description

### PR DESCRIPTION
**Pull request by zklaus**
PR Checklist:

- [x] note any issues closed by this PR with [closing keywords](https://help.github.com/articles/closing-issues-using-keywords)
- [x] if you are adding a new page under `docs/` or `community/`, you have added it to the sidebar in the corresponding `_sidebar.json` file
- [x] put any other relevant information below

This doesn't directly close any issues, but potentially relevant issues and PRs are:

- #1245
- #2000 
- #1665 

🔍 Preview at https://deploy-preview-2167--conda-forge-previews.netlify.app/docs/maintainer/understanding_conda_forge/

### CODE REVIEW

File: docusaurus.config.js [hunk 1]
```diff
@@ -524,6 +524,15 @@ const config = {
...
       },
+      mermaid: {
+        theme: { light: "base", dark: "forest" },
+        options: {
+          themeVariables: {
+            primaryColor: "#b2dfdb",
+            primaryBorderColor: "#00695c",
+          },
+        },
+      },
```
Comment (jaimergp) [inline docusaurus.config.js:535]: Nice. I assume these come from the site theme. A comment might be welcome here.

Also, I noticed in dark mode the diagrams are not really visible, so we might need some adjustments here:

image

Comment (jaimergp) [reply docusaurus.config.js:535]: And while you are at it, we can also change the font family to match the site's. [Apparently these are the options](https://github.com/mermaid-js/mermaid/issues/1614#issuecomment-859462043).

**Comment by jaimergp** [comment]
pre-commit.ci autofix

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
+# The feedstock
+
+A feedstock is the home of one recipe; it may produce more than one conda artifact in the case of a multi-output recipe.
+The feedstock is the stage for most events about a recipe.
```
Comment (ocefpaf) [inline docs/maintainer/understanding_conda_forge/feedstocks.md:4]: I wonder if the use of the word "stage" here, while correct, may confuse some reads. After all, we have a repo named staged-recipes.

BTW, do you believe that for linearity's sake we should start with staged-recipe? I don't have a strong opinion b/c that argument is weak if you are a new contributor to an existing recipe. They the first entry is a feedstock.

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+## Initialization
+
+The feedstock repository is created following the merging in `staged-recipes` (see [Feedstock creation](staged_recipes.md#feedstock-creation) for details on that process).
+
+## Package building diagram
+
+Package building can be triggered by several events, which are described in the next section.
+In all of those cases, the following sequence plays out.
+
+```mermaid
```
Comment (ocefpaf) [inline docs/maintainer/understanding_conda_forge/feedstocks.md:15]: Adding the diagram here just for the sake of visibility. Nice one BTW!

```mermaid
sequenceDiagram
    box github
    participant f as feedstock
    end
    box CI provider
    participant c as CI
    end
    box distribution infrastructure
    participant s as cf-staging channel
    participant cf as conda-forge channel
    participant cdn as CDN
    end
    f->>+c: trigger build
    Note right of c: 1. build<br/>2. validate
    opt if valid and required (4.)
    c->>s: upload packages
    s->>c: report reception success/failure
    Note right of s: 3. validate
    s->>cf: upload packages
    cf->>cdn: upload packages
    end
    c->>-f: report success/failure
```

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+## Events that trigger new builds
+
+Package building is either triggered as part of a PR or on commits to branches in the feedstock repository (which in turn usually stem from the merging of PRs).
+
+:::warning
+_Any_ commit on _any_ branch in the feedstock can lead to the building and publication of a package.
+To avoid the gratuitous publication of improper packages, development branches MUST NOT be added to the feedstock repository.
+Instead, they live in forks of the feedstock repository and the associated work is added only via pull requests.
+:::
```
Comment (ocefpaf) [inline docs/maintainer/understanding_conda_forge/feedstocks.md:57]: Thank you!  I've been meaning to add something like this, so we can link in PRs, forever now but never got it.

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+In addition, other changes to the recipe may be needed, such as updated dependency requirements, a change to the `noarch` status of the package, or an adaptation of the build or test scripts.
+While these last changes often need to be done by the maintainers, conda-forge has sophisticated capabilities to add the initial version update PR.
+
+:::info How does it work?
+
+This happens in two steps in the CI of [`cf-scripts`](https://github.com/regro/cf-scripts).
+
+First, the version information is updated from upstream sources and stored in the [`cf-graph-countyfair` repo](/docs/maintainer/infrastructure/#regrocf-graph-countyfair), more specifically in the `versions` directory tree, nested by hash with one file per package.
+
+```mermaid
```
Comment (ocefpaf) [inline docs/maintainer/understanding_conda_forge/feedstocks.md:103]:
```mermaid
sequenceDiagram
    participant cfs as cf-scripts
    participant gha as github actions
    participant cft as conda_forge_tick
    participant ups as upstream
    participant cfg as cf-graph-countyfair
    loop every hour at :15, :45
        cfs->>gha: bot-versions
        gha->>cft: update-upstream-versions
        cft->>cfg: load package information from `graph.json`
        loop for every package
            cft->>ups: query version
            ups->>cft: return version
            cft->>cfg: write new version
        end
    end
```

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+            ups->>cft: return version
+            cft->>cfg: write new version
+        end
+    end
+```
+
+Second, the main bot CI job, the `bot-bot` action in [`cf-scripts`](/docs/maintainer/infrastructure/#regrocf-scripts) creates PRs for all packages that have a new version available upstream.
+Here is a simplified diagram of how that is done. For the full picture, read [below](#rebuilds-for-migrators).
+
+```mermaid
```
Comment (ocefpaf) [inline docs/maintainer/understanding_conda_forge/feedstocks.md:125]:
```mermaid
sequenceDiagram
    participant cfs as cf-scripts
    participant gha as github actions
    participant cft as conda_forge_tick
    participant fs as feedstock
    participant cfg as cf-graph-countyfair
    loop self renewing
        cfs->>gha: bot-bot
        gha->>cft: auto-tick
        note right of cfg: the graph now contains<br/>the new version information
        cft->>cfg: load package information from `graph.json`
        loop for every package
            opt if new version
                cft->>fs: rewrite recipe and open pr
            end
        end
        gha->>gha: re-trigger bot-bot
    end
```

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+#### Rebuilds for migrators
+
+The version update is one example of a migrator.
+In reality, there are more occasions and reasons to update recipes, for example, to recompile an otherwise unchanged binary program or library to link against a newer version of a dependency or to add support for a new architecture.
+This kind of use case is handled by migrators, which are a general recipe rewriting tool.
+
+A more complete picture of what `auto-tick` does is the following:
+
+```mermaid
```
Comment (ocefpaf) [inline docs/maintainer/understanding_conda_forge/feedstocks.md:160]:
```mermaid
sequenceDiagram
    participant cfs as cf-scripts
    participant gha as github actions
    participant cft as conda_forge_tick
    participant fs as feedstock
    participant cfg as cf-graph-countyfair
    loop self renewing
        cfs->>gha: bot-bot
        gha->>cft: auto-tick
        note right of cfg: the graph now contains<br/>the new version information
        cft->>cfg: load package information from `graph.json`
        loop for every migrator
            create participant mg as migrator
            cft->>mg: run migrator on graph
            note right of mg: 1. filter applicable package to produce effective graph<br/>2. determine migration order<br/>3. run migrator on every possible node
            destroy mg
            mg->>fs: create migration PR
        end
        gha->>cfs: deploy, i.e. commit changed pr information
        gha->>gha: re-trigger bot-bot
    end
```

**Review by ocefpaf — APPROVED** [review]
Love to see this long missing piece in our docs! Thank you for writing it!!

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 8]
```diff
@@ -156,5 +179,4 @@ sequenceDiagram
     end
 ```
 
-Migrators are a powerful mechanism that can do almost arbitrary recipe changes because they are written in Python.
-The current set of migrators can be found [here](https://github.com/regro/cf-scripts/tree/master/conda_forge_tick/migrators).
+Migrators are a powerful mechanism that can do almost arbitrary recipe changes. They are written in Python. The current set of migrators can be found in the [`regro/cf-scripts` repository](https://github.com/regro/cf-scripts/tree/master/conda_forge_tick/migrators).
```
Comment (zklaus) [inline docs/maintainer/understanding_conda_forge/feedstocks.md:184]:
```suggestion
Migrators are a powerful mechanism that can do almost arbitrary recipe changes.
They are written in Python and the current set of migrators can be found in the [`regro/cf-scripts` repository](https://github.com/regro/cf-scripts/tree/master/conda_forge_tick/migrators).
```

File: docs/maintainer/understanding_conda_forge/life_cycle.md [hunk 9]
```diff
@@ -74,7 +74,7 @@ If successful, the artifacts will then be copied to the actual `conda-forge` cha
 
 At this point, the channel server will process the contents of the new packages to retrieve their metadata and update the repodata files.
 On the next CDN sync cycle, the artifacts will be distributed to the delivery network for faster access.
-Validation and CDN sync usually take less than 15 minutes after the CI is passing on `main`. From this moment on, users can install the new packages from the CLI.
+Validation and CDN sync usually takes around 15 minutes after the CI is passing on `main`. From this moment on, users can install the new packages from the CLI.
```
Comment (zklaus) [inline docs/maintainer/understanding_conda_forge/life_cycle.md:77]:
```suggestion
Validation and CDN sync usually take around 15 minutes after the CI is passing on `main`.
From this moment on, users can install the new packages from the CLI.
```

File: docs/maintainer/understanding_conda_forge/life_cycle.md [hunk 10]
```diff
@@ -91,12 +91,12 @@ However, they are still available via direct URL access.
 This allows organizations to retire packages from normal, solver-driven installs without compromising the reproducibility offered by lock files.
 
 Finally, a project might have reached a status where no further updates are needed or expected (e.g. it has been superseded by a new project).
-If the maintainers want to, these feedstocks can be archived and marked as read-only.
+If the maintainers want to, these feedstocks can be archived and marked as read-only. This is also done via [`admin-requests`](/docs/maintainer/infrastructure#admin-requests).
```
Comment (zklaus) [inline docs/maintainer/understanding_conda_forge/life_cycle.md:94]:
```suggestion
If the maintainers want to, these feedstocks can be archived and marked as read-only, which is also done via [`admin-requests`](/docs/maintainer/infrastructure#admin-requests).
```

File: docs/maintainer/understanding_conda_forge/feedstocks.md (hunk 2 above)
Comment (jaimergp) [reply docs/maintainer/understanding_conda_forge/feedstocks.md:4]: The suggested reading order is the one in sidebar, so yes, staged-recipes goes first:

image

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 11]
```diff
@@ -41,39 +39,63 @@ sequenceDiagram
...
+
+### Manually submitted PRs
+
+These PRs are not automated. Any Github user can open a new PR by forking the feedstock, creating a fresh branch from `main` and adding the necessary commits to achieve the intended change (e.g. building a new version or adding a new platform).
+
+It is up to the feedstock maintainers and/or the conda-forge/core team to review, approve and merge the PR, or to reject it and close it.
+
+During the lifetime of the PR, some automated operations will take place:
+
+- The linter will scan the state of the recipe to require changes and suggest improvements. If not fulfilled, this will result in a failed CI run. Linter failures must **not** be ignored without explicit core approval.
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply docs/maintainer/understanding_conda_forge/feedstocks.md:80]: I wanted to emphasize the NOT here as done in CEPs and so on. No strong feelings, and we don't have style guide in this regard either. Aesthetically speaking the italics in this font are too slanted for my taste so I'm actually looking at other variants of the Inter font to "fix it".

File: docs/maintainer/understanding_conda_forge/feedstocks.md [hunk 12]
```diff
@@ -41,39 +39,63 @@ sequenceDiagram
...
-2. The validation checks that the artifacts produced during the build are permissible for this feedstock by consulting [the `feedstock-outputs` repo](/docs/maintainer/infrastructure/#feedstock-outputs)
+1. For Linux and `noarch` packages, the build itself is carried out in a Docker container on the CI. On macOS and Windows, the CI runner system image is used after tuning it a bit.
+2. The validation checks that the artifacts produced during the build are permissible for this feedstock by consulting [the `feedstock-outputs` repository](/docs/maintainer/infrastructure/#feedstock-outputs).
 3. The server-side validation is essentially the same as 2. It is repeated to guard against potential intentional or unintentional interference at the feedstock level, which is easier to access.
-4. The upload is only triggered if the validation (2.) was successful. Additionally, it is only performed under certain conditions, for example for commits in `main`, but _not_ in PRs.
+4. The upload is only triggered if the validation (3.) was successful. Additionally, it is only performed under certain conditions, for example for commits in `main`, but _not_ in PRs.
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply docs/maintainer/understanding_conda_forge/feedstocks.md:45]: If the server-side validation doesn't work then there's no (full) upload, right? As in it won't ever make it to conda-forge (only cf-staging).

**Comment by jaimergp** [comment]
pre-commit.ci autofix

File: docs/maintainer/understanding_conda_forge/feedstocks.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (zklaus) [reply docs/maintainer/understanding_conda_forge/feedstocks.md:45]: True. But if 2. fails, there won't be either. In that case, it is much clearer to me what to do (fix the problem in the PR). If 2. succeeds, but 3. fails, it is less clear how to proceed. Can we give guidance for that?

[the remark this answers is not in the document]
Comment (jaimergp) [reply docs/maintainer/understanding_conda_forge/feedstocks.md:45]: What about this then?

```suggestion
4. The upload is only triggered if the validation steps (both 2. and 3.) were successful. Additionally, it is only performed under certain conditions, for example for commits in `main`, but _not_ in PRs.
```

We can discuss validation in its own page once we document that.

[the remark this answers is not in the document]
Comment (zklaus) [reply docs/maintainer/understanding_conda_forge/feedstocks.md]: @jaimergp and I discussed this offline and I updated the description accordingly.

**Comment by zklaus** [comment]
I think all comments were addressed.

**Comment by jaimergp** [comment]
Let's merge! Further refinements, if needed, can happen in additional PRs.

**Pull request merged**